spec/ameba/rule/lint/unused_comparison_spec.cr

```diff
@@ -30,6 +30,21 @@ module Ameba::Rule::Lint
         CRYSTAL
     end
 
+    it "passes for comparisons inside '||' and '&&' where the other arg is a call" do
+      expect_no_issues subject, <<-CRYSTAL
+          IO.copy(in_var, out_var, len) == len || raise IO::EOFError.new
+        CRYSTAL
+    end
+
+    it "passes if a regex literal is part of a === or =~ comparison" do
+      expect_no_issues subject, <<-CRYSTAL
+        /f(o+)(bar?)/ === "fooba"
+        puts $~
+        "foobar" =~ /(o+)ba(r?)/
+        puts $1
+      CRYSTAL
+    end
```

Comment:
This doesn't make much sense, since regex literals always have side-effects (setting `$N` variables and alike).

I think `RegexLiteral`-support should be removed altogether.

Reply:
From what I can see looking at the stdlib, `$~` is only set when using these two operators (or the `match` methods), and `$1` is just a shorthand for `$~[1]`.

Reply:
Ameba probably should not make any assumptions about implementations in stdlib. And without semantic analysis there's no way to know whether any call even targets the stdlib implementation.

Reply:
That's true, so for unused comparison we should check to see if either the obj or the arg is a regex, and not apply it if so. But in the case of unused literal, there isn't a situation in which a regex literal by itself in a void context will have side effects, right?

Reply:
> That's true, so for unused comparison we should check to see if either the obj or the arg is a regex, and not apply it if so.

That would still cause false positive for non-literal instances of `Regex`, such as `re = /(.)oo/; re === "foo"`.
Without semantical analysis this cannot be solved correctly.
The best course of action is probably to drop `===` and `~=` from the list of tested comparison operators.

Other operators should generally not be affected, at least not in stdlib. The semantics defined in the language itself don't suggest any realistic use case with side effects, so they should probably be acceptable.

Reply:
> But in the case of unused literal, there isn't a situation in which a regex literal by itself in a void context will have side effects, right?

Probably not. A literal expands to a call to `Regex.literal` which compiles the regex. That *might* cause some side effect inside the regex engine, but that shouldn't be critical.